traffic_ops/app/db/admin.pl

```diff
@@ -14,261 +14,8 @@
-	my $open = $db_connection->{open};
-
-	# Goose requires the 'open' line in the dbconf file to be a scalar.
-	# example:
-	#		open: host=127.0.0.1 port=5432 user=to_user password=twelve dbname=to_development sslmode=disable
-	# We need access to these values for db connections so I am manipulating the 'open'
-	# line so that it can be loaded into a hash.
-	$open = join "\n", map { s/=/ : /; $_ } split " ", $open;
-	my $hash = Load $open;
-
-	$host_ip     = $hash->{host};
-	$host_port   = $hash->{port};
-	$db_user = $hash->{user};
-	$db_password = $hash->{password};
-	$db_name     = $hash->{dbname};
-}
-
-sub migrate {
-	my ($command) = @_;
-
-	print "Migrating database...\n";
-	if ( system("goose --env=$environment $command") != 0 ) {
-		die "Can't run goose\n";
-	}
-}
-
-sub seed {
-	print "Seeding database w/ required data.\n";
-	local $ENV{PGPASSWORD} = $db_password;
-	if ( system("psql -h $host_ip -p $host_port -d $db_name -U $db_user -e -v ON_ERROR_STOP=1 < db/seeds.sql") != 0 ) {
-		die "Can't seed database w/ required data\n";
-	}
-}
-
-sub patches {
-	print "Patching database with required data fixes.\n";
-	local $ENV{PGPASSWORD} = $db_password;
-	if ( system("psql -h $host_ip -p $host_port -d $db_name -U $db_user -e -v ON_ERROR_STOP=1 < db/patches.sql") != 0 ) {
-		die "Can't patch database w/ required data\n";
-	}
-}
-
-sub load_schema {
-	print "Creating database tables.\n";
-	local $ENV{PGPASSWORD} = $db_password;
-	if ( system("psql -h $host_ip -p $host_port -d $db_name -U $db_user -e -v ON_ERROR_STOP=1 < db/create_tables.sql") != 0 ) {
-		die "Can't create database tables\n";
-	}
-}
-
-sub dropdb {
-	print "Dropping database: $db_name\n";
-	if ( system("dropdb -h $host_ip -p $host_port -U $db_super_user -e --if-exists $db_name;") != 0 ) {
-		die "Can't drop db $db_name\n";
-	}
-}
-
-sub createdb {
-	my $db_exists = `psql -h $host_ip -U $db_super_user -p $host_port -tAc "SELECT 1 FROM pg_database WHERE datname='$db_name'"`;
-	if ($db_exists) {
-		print "Database $db_name already exists\n";
-		return;
-	}
-	my $cmd = "createdb -h $host_ip -p $host_port -U $db_super_user -e --owner $db_user $db_name;";
-	if ( system($cmd) != 0 ) {
-		die "Can't create db $db_name\n";
-	}
-
-}
-
-sub create_user {
-	print "Creating user: $db_user\n";
-	my $user_exists = `psql -h $host_ip -p $host_port -U $db_super_user -tAc "SELECT 1 FROM pg_roles WHERE rolname='$db_user'"`;
-
-	if (!$user_exists) {
-		my $cmd = "CREATE USER $db_user WITH LOGIN ENCRYPTED PASSWORD '$db_password'";
-		if ( system(qq{psql -h $host_ip -p $host_port -U $db_super_user -etAc "$cmd"}) != 0 ) {
-			die "Can't create user $db_user\n";
-		}
-	}
-}
-
-sub drop_user {
-	if ( system("dropuser -h $host_ip -p $host_port -U $db_super_user-i -e $db_user;") != 0 ) {
-		die "Can't drop user $db_user\n";
-	}
-}
-
-sub show_users {
-	if ( system("psql -h $host_ip -p $host_port -U $db_super_user -ec '\\du';") != 0 ) {
-		die "Can't show users";
-	}
-}
-
-sub reverse_schema {
```

Comment:
Shoot, I just remembered that this function is still needed as long as we have Perl TO API endpoints, and the admin binary actually shells out this command to this admin.pl script. So this function should probably be pulled out into its own little Perl script that the admin binary can shell that command out to. Then admin.pl can just be as you have it now. Otherwise, in this PR if you try running this command, I think it will just create an endless cycle due to the circular dependency.

Reply:
shouldn't that be implemented in the binary?

Reply:
I'm concerned though - because this needs to exist as long as the Perl back-end exists, and because we deprecated the script, it sort of seems like we're now locked in to a release schedule that demands that 4.0 not be released until the entire Perl TO codebase can be removed.

Reply:
Can we put `reverse_schema` in its own script file, `admin_reverse_schema.pl`, with only that function, and make Go execute that script?

It's not ideal, but the only dependencies it will have is DBIx, which fortunately is in a CentOS package, so it shouldn't need CPan or Carton.

Reply:
We could, but idk how much that really buys us. It also has a dependency on `Schema` btw, but I think that's maybe a standard library thing? Not sure. Leaving it in `admin.pl` means adding two dependencies: `GetOpt::Long` (standard?) and `YAML`, which I don't personally think of as a huge deal, but it is two places to maintain the command line argument format. On the other hand separating it means we now have three programs to maintain for the `admin` tool, at least until `admin.pl` vanishes. It's also functionality specific to `admin.pl` since it only deals with Perl data representation and should vanish with the need for that script

In summary, I'm sort of torn on the idea of putting that in a new script, and I'm leaning towards saying that that's something that should maybe take place once `admin.pl` actually gets removed (assuming the Perl back-end still exists at that point).

Reply:
>It also has a dependency on Schema 
`Schema` is DBIx.

>two dependencies: GetOpt::Long (standard?) and YAML

>two places to maintain the command line argument format

I'm not suggesting maintaining the argument format. Make the script take the arguments it needs (environment, user, pass, host, port), and make Go pass them. No YAML or GetOpt.

Reply:
I think you misunderstood, those dependencies and that maintenance *only apply* if we're leaving it in `admin.pl`, not if we put it in a separate script as you suggest.

Reply:
We shouldn't really be worried about which dependencies the new reverse_schema.pl script will require, since as long as we have Perl endpoints we will still have the Perl dependencies required for reverse_schema. For now, yes, we will end up with 3 programs to maintain until we get rid of admin.pl, but that shouldn't really be a concern either because reverse_schema.pl will only contain that required functionality and won't require any further updates until it's eventually removed with TO-Perl. And with admin.pl just being a thin Perl wrapper over the admin.go binary, there really won't be much to maintain there either.

Reply:
the alternative I'm suggesting is keep that one function in `admin.pl` and remove everything else. It's not as clean as making a new script, I guess, but that's not super clean either. I suppose I just don't see the advantage of moving this to a new file until we're prepared to get rid of `admin.pl` in 4.0.

Reply:
As long as `admin.pl` is kept as an empty wrapper, having real functionality called by `admin.go` creates a circular dependency. It can work, but circular scripts are perilous. Much better, simpler, safer, easier to understand, to make a third script to avoid the cycle.

Reply:
Well I was going to have `admin.pl` intercept the `reverse_schema` functionality such that it never calls `admin` in that case. No circularity required.

Reply:
But `admin.go` has to call `admin.pl` for that function as well. The scripts are still calling each-other. Even if the code execution isn't circular, the scripts are. That's confusing and dangerous.

Reply:
Well in any case I don't feel strongly about it. I can separate it into its own script.

Reply:
The `get_dbinfo` function checks the environment variables for MOJO_MODE and uses that to determine which database.conf to read the config from. By default it will use conf/development/database.conf. That comes from traffic_ops/app/lib/Schema.pm I believe.

I'll pull this down and take a spin.